File: DataStructureAndAlgo/unionfind.py
#!/usr/bin/env python3


class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
        self.components = n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x, y):
        rx, ry = self.find(x), self.find(y)
        if rx == ry:
            return False

        if self.size[ry] > self.size[rx]:
            rx, ry = ry, rx

        self.size[rx] += self.size[ry]
        self.parent[ry] = rx
        self.components -= 1
        return True

    def connected(self, x, y):
        return self.find(x) == self.find(y)

    def get_size(self, x):
        return self.size[self.find(x)]

    def count_components(self):
        return self.components

File: DataStructureAndAlgo/UnionFind.swift
class UnionFind {
    private var parent: [Int]
    private var size: [Int]

    init(_ n: Int) {
        parent = Array(0 ..< n)
        size = Array(repeating: 1, count: n)
    }

    func find(_ x: Int) -> Int {
        if parent[x] != x {
            parent[x] = find(parent[x])
        }
        return parent[x]
    }

    func union(_ x: Int, _ y: Int) -> Bool {
        var rx = find(x), ry = find(y)
        guard rx != ry else { return false }

        if size[ry] > size[rx] { swap(&rx, &ry) }
        parent[ry] = rx
        size[rx] += size[ry]
        return true
    }

    func connected(_ x: Int, _ y: Int) -> Bool {
        return find(x) == find(y)
    }
}
